core/platform/auth/firebase_auth_services.py

```diff
@@ -0,0 +1,203 @@
+        third-party application to obtain limited access to an HTTP service on
+        behalf of a resource owner.
+    Claim:
+        A piece of information about a user (name, address, phone number, etc.)
+        that has been encrypted and digitally signed.
+    JSON Web Token (JWT):
+        A compact and URL-safe protocol primarily designed to send Claims
+        between two parties. Claims are organized into JSON objects mapping
+        "Claim Names" to "Claim Values".
+    Identity provider:
+        An entity that creates, maintains, and manages identity information and
+        provides authentication services. Such services rely on JWTs to send
+        identity information. Examples of identity providers include: Google,
+        Facebook, Email verification links, and Text message SMS codes.
+    Subject Identifier:
+        A Claim that can uniquely identify a user. It is locally unique and
+        never reassigned with respect to the provider who issued it.
+        For example: `24400320` or `AItOawmwtWwcT0k51BayewNvutrJUqsvl6qs7A4`.
+"""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+import logging
+
+from core.platform import models
+import python_utils
+
+import firebase_admin
+from firebase_admin import auth as firebase_auth
+
+(auth_models,) = models.Registry.import_models([models.NAMES.auth])
+
+
+def _ensure_firebase_is_initialized():
+    """Initializes the Firebase Admin SDK."""
+    try:
+        firebase_admin.get_app()
+    except Exception:
+        firebase_admin.initialize_app()
+
+
+def authenticate_request(request):
+    """Authenticates the request and returns the user who authorized it, if any.
+
+    Oppia follows the OAuth Bearer authentication scheme.
+
+    Bearer authentication (a.k.a. token authentication) is an HTTP
+    authentication scheme based on "bearer tokens", an encrypted JWT generated
+    by a trusted identity provider in response to login requests.
+
+    The name "Bearer authentication" can be understood as: "give access to the
+    bearer of this token." These tokens _must_ be sent in the `Authorization`
+    header of HTTP requests and _must_ have the format: `Bearer <token>`.
+
+    Oppia specifically expects the token to have a Subject Identifier Claim for
+    the user (Claim Name: 'sub').
+
+    To learn more:
+        HTTP authentication schemes:
+            https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication
+        OAuth 2.0 Bearer authentication scheme:
+            https://oauth.net/2/bearer-tokens/
+
+    Args:
+        request: webapp2.Request. The HTTP request to inspect.
+
+    Returns:
+        str|None. The ID of the user that authorized the request. If the request
+        could not be authenticated, then returns None instead.
+    """
+    try:
+        _ensure_firebase_is_initialized()
+    except Exception as e:
```

Comment:
The same comment as in `_ensure_firebase_is_initialized` above.

Reply:
ditto below

Reply:
Done.